File: LC/53/53.java
class Solution {
    public int maxSubArray(int[] nums) {
        if (nums == null) {
            return 0;
        }
        
        int start = 0;
        int sum = 0;
        int maxInt = Integer.MIN_VALUE;
        for (int end = 0; end < nums.length; end++) {
            sum += nums[end];
            while (sum < nums[end]) {
                sum -= nums[start];
                start++;
            }
            maxInt = Math.max(maxInt, sum);
        }
        
        return maxInt;
    }
}

File: LC/53/53.ts
function maxSubArray(nums: number[]): number {
    if (nums.length == 0) return 0;
    
    let l = 0;
    let maxSum = nums[0],
        sum = nums[0];
    for (let r = 1; r < nums.length; r++) {
        sum += nums[r];
        while (l < r && nums[r] > sum) sum -= nums[l++];
        maxSum = Math.max(maxSum, sum);
    }
    return maxSum;
};
